File: ios/Hapi/Features/Sessions/AgentFlavorIcon.swift
import SwiftUI

/// Brand logo per agent flavor — port of `web/src/components/AgentFlavorIcon.tsx`
/// (brand SVGs via @lobehub/icons, the same source the web ships; assets live
/// in `Assets.xcassets/AgentIcons`).
///
/// Color variants (agy/claude/codex/gemini) carry `original` rendering intent
/// so the literal brand fills stay visible on light and dark — the reason the
/// web picked Color for them. Mono variants (cursor/grok/kimi/opencode/pi)
/// are template images and inherit the environment foreground style (the
/// web's currentColor); copilot mirrors the web's fixed GitHub-mark tint
/// (#24292F light / #E6EDF3 dark). Unknown flavors fall back to the web's
/// "Un" badge.
struct AgentFlavorIconView: View {
    let flavor: String?
    var size: CGFloat = 16

    @Environment(\.colorScheme) private var colorScheme

    /// Flavors with a bundled `AgentIcons/<flavor>` image set.
    private static let bundledFlavors: Set<String> = [
        "agy", "claude", "codex", "copilot", "cursor",
        "gemini", "grok", "kimi", "opencode", "pi",
    ]

    /// Asset name for a raw `metadata.flavor` value; nil when no icon ships.
    static func assetName(forFlavor flavor: String?) -> String? {
        guard let flavor else { return nil }
        let normalized = flavor.trimmingCharacters(in: .whitespaces).lowercased()
        return bundledFlavors.contains(normalized) ? normalized : nil
    }

    var body: some View {
        content
            .frame(width: size, height: size)
            .accessibilityHidden(true)
    }

    @ViewBuilder
    private var content: some View {
        if let name = Self.assetName(forFlavor: flavor) {
            let icon = Image(name)
                .resizable()
                .scaledToFit()
            if name == "copilot" {
                icon.foregroundStyle(colorScheme == .dark
                    ? Color(red: 230 / 255, green: 237 / 255, blue: 243 / 255)
                    : Color(red: 36 / 255, green: 41 / 255, blue: 47 / 255))
            } else {
                icon
            }
        } else {
            Text("Un")
                .font(.system(size: 8, weight: .semibold))
                .foregroundStyle(.secondary)
                .frame(maxWidth: .infinity, maxHeight: .infinity)
                .background(Color.secondary.opacity(0.15), in: RoundedRectangle(cornerRadius: 3))
        }
    }
}

#Preview("Agent icons") {
    AgentFlavorIconPreviewGrid()
}

#Preview("Agent icons · dark") {
    AgentFlavorIconPreviewGrid()
        .preferredColorScheme(.dark)
}

private struct AgentFlavorIconPreviewGrid: View {
    private let flavors: [String?] = [
        "agy", "claude", "codex", "copilot", "cursor",
        "gemini", "grok", "kimi", "opencode", "pi", "mystery",
    ]

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            ForEach(flavors, id: \.self) { flavor in
                HStack(spacing: 8) {
                    AgentFlavorIconView(flavor: flavor)
                    Text(flavor ?? "nil")
                        .font(.footnote)
                        .foregroundStyle(.secondary)
                }
            }
        }
        .padding()
    }
}

File: android/app/src/main/kotlin/app/hapi/companion/ui/components/AgentFlavorIcon.kt
package app.hapi.companion.ui.components

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.hapi.companion.R
import app.hapi.companion.ui.theme.HapiTheme
import app.hapi.companion.ui.theme.hapi

/**
 * Brand logo per agent flavor — port of `web/src/components/AgentFlavorIcon.tsx`
 * (brand SVGs via @lobehub/icons, the same source the web ships).
 *
 * Color variants (agy/claude/codex/gemini) keep their literal brand fills and
 * render untinted so they stay visible on both light and dark surfaces — the
 * reason the web picked Color for them. Mono variants (cursor/grok/kimi/
 * opencode/pi) are currentColor upstream and tint with [LocalContentColor];
 * copilot mirrors the web's fixed GitHub-mark tint (#24292F light /
 * #E6EDF3 dark). Unknown flavors fall back to the web's "Un" badge.
 */
@Composable
fun AgentFlavorIcon(flavor: String?, modifier: Modifier = Modifier.size(16.dp)) {
    when (val normalized = flavor?.trim()?.lowercase().orEmpty()) {
        in COLOR_FLAVOR_ICONS -> Image(
            painter = painterResource(COLOR_FLAVOR_ICONS.getValue(normalized)),
            contentDescription = null,
            modifier = modifier,
        )

        in MONO_FLAVOR_ICONS -> Icon(
            painter = painterResource(MONO_FLAVOR_ICONS.getValue(normalized)),
            contentDescription = null,
            modifier = modifier,
            tint = LocalContentColor.current,
        )

        "copilot" -> Icon(
            painter = painterResource(R.drawable.ic_agent_copilot),
            contentDescription = null,
            modifier = modifier,
            tint = if (MaterialTheme.hapi.isDark) Color(0xFFE6EDF3) else Color(0xFF24292F),
        )

        else -> Box(
            modifier = modifier.background(
                color = MaterialTheme.colorScheme.surfaceContainerHigh,
                shape = RoundedCornerShape(3.dp),
            ),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                text = "Un",
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                fontSize = 8.sp,
                lineHeight = 8.sp,
                fontWeight = FontWeight.SemiBold,
                maxLines = 1,
                softWrap = false,
            )
        }
    }
}

private val COLOR_FLAVOR_ICONS: Map<String, Int> = mapOf(
    "agy" to R.drawable.ic_agent_agy,
    "claude" to R.drawable.ic_agent_claude,
    "codex" to R.drawable.ic_agent_codex,
    "gemini" to R.drawable.ic_agent_gemini,
)

private val MONO_FLAVOR_ICONS: Map<String, Int> = mapOf(
    "cursor" to R.drawable.ic_agent_cursor,
    "grok" to R.drawable.ic_agent_grok,
    "kimi" to R.drawable.ic_agent_kimi,
    "opencode" to R.drawable.ic_agent_opencode,
    "pi" to R.drawable.ic_agent_pi,
)

// -------------------------------------------------------------- preview --

@Composable
private fun AgentFlavorIconStrip() {
    Column(
        modifier = Modifier.padding(12.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        val flavors = listOf(
            "agy", "claude", "codex", "copilot", "cursor",
            "gemini", "grok", "kimi", "opencode", "pi", "mystery",
        )
        flavors.chunked(6).forEach { chunk ->
            Row(
                horizontalArrangement = Arrangement.spacedBy(10.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                chunk.forEach { flavor -> AgentFlavorIcon(flavor) }
            }
        }
    }
}

@Preview(showBackground = true, name = "Agent icons · light")
@Composable
private fun AgentFlavorIconPreviewLight() {
    HapiTheme(darkTheme = false, dynamicColor = false) {
        Surface { AgentFlavorIconStrip() }
    }
}

@Preview(showBackground = true, name = "Agent icons · dark")
@Composable
private fun AgentFlavorIconPreviewDark() {
    HapiTheme(darkTheme = true, dynamicColor = false) {
        Surface { AgentFlavorIconStrip() }
    }
}
